File: packages/agent-runtime/src/low-priority.ts
import { existsSync } from 'node:fs';

export interface LowPriorityCommand {
  bin: string;
  args: string[];
}

/** Default child `nice` increment — a courteous but not starved background priority. */
const DEFAULT_NICENESS = 10;

/**
 * Wrap an agent/helper child in OS CPU/I/O priority controls (#456 A6, #528
 * Phase D). Per-platform mapping of what actually applies:
 *   - Linux: `nice` (CPU) + best-effort idle-class `ionice` (I/O); a child
 *     could additionally opt into `SCHED_IDLE`, but that needs a syscall wrapper
 *     we don't ship, so `nice`/`ionice` are the portable floor.
 *   - macOS: `nice` (CPU). App Nap applies to the child on its own when it's
 *     backgrounded and not holding power assertions — no flag to pass.
 *   - Windows: no wrapper — Node has no creation-priority flag, so correctness
 *     must not depend on one. `CENTRAID_CHILD_PRIORITY=normal` is the documented
 *     passthrough to disable wrapping everywhere; a future EcoQoS
 *     (PROCESS_POWER_THROTTLING) opt-in is deferred.
 *
 * worker_threads DEFERRED: Node cannot set per-thread OS priority portably
 * (no cross-platform tid-based `setPriority`), so pool threads inherit the
 * process priority — do not hack a tid lookup to force it.
 *
 * `niceness` overrides the default increment (still ignored on Windows and
 * when `CENTRAID_CHILD_PRIORITY=normal`).
 */
export function lowPriorityCommand(
  bin: string,
  args: readonly string[],
  options: {
    platform?: NodeJS.Platform;
    exists?: (file: string) => boolean;
    niceness?: number;
  } = {},
): LowPriorityCommand {
  const platform = options.platform ?? process.platform;
  const exists = options.exists ?? existsSync;
  const niceness = options.niceness ?? DEFAULT_NICENESS;
  if (platform === 'win32' || process.env.CENTRAID_CHILD_PRIORITY === 'normal') {
    return { bin, args: [...args] };
  }
  const nice = exists('/usr/bin/nice') ? '/usr/bin/nice' : 'nice';
  const niceArgs = ['-n', String(niceness), '--', bin, ...args];
  if (platform !== 'linux') return { bin: nice, args: niceArgs };
  const ionice = ['/usr/bin/ionice', '/bin/ionice'].find(exists);
  return ionice
    ? { bin: ionice, args: ['-c', '2', '-n', '7', nice, ...niceArgs] }
    : { bin: nice, args: niceArgs };
}

File: crates/assist/src/low_priority.rs
//! CPU and I/O priority for a harness child (#456/#528; #1020, D-1020-AS1).
//!
//! A harness turn is a long-running subprocess that will happily saturate four
//! cores and the disk queue while the member is trying to scroll a list. So it
//! is launched at low priority — and *launched* is the operative word: the
//! niceness is applied by wrapping the command in `nice` (and `ionice` on
//! Linux) rather than by calling `setpriority` after the fact, because the
//! window between spawn and adjustment is exactly when a cold start does its
//! worst.
//!
//! `low-priority.ts:10`–`:12`: **worker_threads is deferred — do not hack a tid
//! lookup.** The Rust port inherits the same non-goal for its own threads: this
//! module lowers *child processes*, and nothing here reaches into a thread id.
//!
//! Windows gets nothing, deliberately: there is no `nice`, and the
//! `SetPriorityClass` route needs a handle to a process that is already
//! running, which is the race this design avoids.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

const DEFAULT_NICENESS: i32 = 10;

/// A command, after the priority wrapper has been applied.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Command {
    pub program: PathBuf,
    pub args: Vec<String>,
}

/// What the wrapper is allowed to assume about the host.
#[derive(Debug, Clone)]
pub struct Host {
    /// `"linux"`, `"macos"`, `"windows"` — the three cases that differ.
    pub platform: &'static str,
    pub niceness: i32,
    /// Whether a path exists. Injected so the decision is testable without
    /// depending on what happens to be installed on the runner.
    pub exists: fn(&Path) -> bool,
    /// `CENTRAID_CHILD_PRIORITY=normal` opts a host out entirely — the escape
    /// hatch for a machine where `nice` is denied by a sandbox.
    pub opt_out: bool,
}

impl Host {
    /// The real host.
    #[must_use]
    pub fn current(env: &BTreeMap<String, String>) -> Self {
        Self {
            platform: if cfg!(target_os = "windows") {
                "windows"
            } else if cfg!(target_os = "linux") {
                "linux"
            } else {
                "other"
            },
            niceness: DEFAULT_NICENESS,
            exists: |path| path.exists(),
            opt_out: env.get("CENTRAID_CHILD_PRIORITY").map(String::as_str) == Some("normal"),
        }
    }
}

/// Wrap a command so the child runs at low CPU and (on Linux) low I/O
/// priority.
#[must_use]
pub fn low_priority_command(program: &Path, args: &[String], host: &Host) -> Command {
    if host.platform == "windows" || host.opt_out {
        return Command {
            program: program.to_path_buf(),
            args: args.to_vec(),
        };
    }
    let nice = if (host.exists)(Path::new("/usr/bin/nice")) {
        PathBuf::from("/usr/bin/nice")
    } else {
        PathBuf::from("nice")
    };
    // `--` matters: without it a harness argument that starts with `-` is read
    // by `nice` as one of its own options.
    let mut nice_args = vec![
        "-n".to_owned(),
        host.niceness.to_string(),
        "--".to_owned(),
        program.display().to_string(),
    ];
    nice_args.extend(args.iter().cloned());

    if host.platform != "linux" {
        return Command {
            program: nice,
            args: nice_args,
        };
    }
    let ionice = ["/usr/bin/ionice", "/bin/ionice"]
        .into_iter()
        .map(PathBuf::from)
        .find(|candidate| (host.exists)(candidate));
    match ionice {
        // Best-effort class, lowest priority within it: the harness yields the
        // disk to whatever the member is doing, and still makes progress.
        Some(ionice) => {
            let mut args_with_nice = vec![
                "-c".to_owned(),
                "2".to_owned(),
                "-n".to_owned(),
                "7".to_owned(),
                nice.display().to_string(),
            ];
            args_with_nice.extend(nice_args);
            Command {
                program: ionice,
                args: args_with_nice,
            }
        }
        None => Command {
            program: nice,
            args: nice_args,
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn host(platform: &'static str, exists: fn(&Path) -> bool) -> Host {
        Host {
            platform,
            niceness: 10,
            exists,
            opt_out: false,
        }
    }

    fn all(_path: &Path) -> bool {
        true
    }
    fn none(_path: &Path) -> bool {
        false
    }

    #[test]
    fn linux_wraps_in_ionice_then_nice() {
        let command = low_priority_command(
            Path::new("codex"),
            &["--acp".to_owned()],
            &host("linux", all),
        );
        assert_eq!(command.program, PathBuf::from("/usr/bin/ionice"));
        assert_eq!(
            command.args,
            [
                "-c",
                "2",
                "-n",
                "7",
                "/usr/bin/nice",
                "-n",
                "10",
                "--",
                "codex",
                "--acp"
            ]
        );
    }

    #[test]
    fn macos_wraps_in_nice_only() {
        let command = low_priority_command(Path::new("codex"), &[], &host("other", all));
        assert_eq!(command.program, PathBuf::from("/usr/bin/nice"));
        assert_eq!(command.args, ["-n", "10", "--", "codex"]);
    }

    #[test]
    fn windows_is_left_alone() {
        let command = low_priority_command(
            Path::new("codex.exe"),
            &["--acp".to_owned()],
            &host("windows", all),
        );
        assert_eq!(command.program, PathBuf::from("codex.exe"));
        assert_eq!(command.args, ["--acp"]);
    }

    #[test]
    fn a_host_with_neither_helper_falls_back_to_bare_nice() {
        let command = low_priority_command(Path::new("codex"), &[], &host("linux", none));
        assert_eq!(command.program, PathBuf::from("nice"));
    }

    #[test]
    fn the_opt_out_returns_the_command_untouched() {
        let mut opted = host("linux", all);
        opted.opt_out = true;
        let command = low_priority_command(Path::new("codex"), &[], &opted);
        assert_eq!(command.program, PathBuf::from("codex"));
        assert!(command.args.is_empty());
    }
}
